# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1770855621 -39600
# Parent  76c3e516e1cf7936251380d1a8bb5cd10404f989
Misc: Introduce pref 'mail.folderpane.unified_include_subfolders' to unconditionally include all subfolders in unified folders.

Up to version 102, a unified folder of a type contained all folders of that type from all accounts, regardless of whether
they were part of the search, and also their subfolders. For example, the unified inbox contained all inboxes and their subfolders
even if those weren't part of the search folders of the unified/virtual folder.

From version 115, only folders nominated as search folders are displayed in the unified folders.

If may be desired to show subfolders, although they are not included as search folders.

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -326,8 +326,9 @@ pref("mail.folderpane.sizeUnits", "");
 // Summarize messages count and size of subfolders into a collapsed parent?
 // Allowed values: true/false
 pref("mail.folderpane.sumSubfolders", true);
 pref("mail.folderpane.expand_delay", 750);
+pref("mail.folderpane.unified_include_subfolders", false);
 
 // target folder URI used for the last move or copy
 pref("mail.last_msg_movecopy_target_uri", "");
 // last move or copy operation was a move
diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -1130,8 +1130,9 @@ var folderPane = {
                   searchFolder
                 );
               }
             }
+            this.includeSubfolders(wrappedFolder);
           } catch (ex) {
             console.error(
               new Error(
                 `Unable to access the search folders of ${folder.URI}`,
@@ -1142,8 +1143,33 @@ var folderPane = {
         }
         MailServices.accounts.saveVirtualFolders();
       },
 
+      includeSubfolders(wrappedFolder) {
+        if (
+          !Services.prefs.getBoolPref(
+            "mail.folderpane.unified_include_subfolders",
+            false
+          )
+        ) {
+          return;
+        }
+        for (const searchFolder of wrappedFolder.searchFolders) {
+          for (const sf of searchFolder.subFolders) {
+            const existingRow = folderPane.getRowForFolder(sf, this.name);
+            if (!existingRow) {
+              const parentRow = folderPane.getRowForFolder(
+                searchFolder,
+                this.name
+              );
+              parentRow.insertChildInOrder(
+                folderPane._createFolderRow(this.name, sf)
+              );
+            }
+          }
+        }
+      },
+
       regenerateMode() {
         if (this._smartMailbox) {
           SmartMailboxUtils.removeAll(true);
         }
@@ -1248,8 +1274,10 @@ var folderPane = {
             searchFolder
           );
         }
 
+        this.includeSubfolders(wrappedFolder);
+
         // For any rows we removed, check they are added back to the tree.
         for (const server of serversToCheck) {
           this.initServer(server);
         }
